Intermittent failure in pytest - `test_jit_sort_joins`

### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.


### Issue description

I can reliably reproduce this when running 12 threads. Succeeds on 6 threads.

I'll take a look at this later, just reporting this so I don't forget.

```
________________________________________________________________ test_jit_sort_joins _________________________________________________________________
[gw7] linux -- Python 3.11.0 /home/stijn/code/polars/py-polars/venv/bin/python

    @typing.no_type_check
    def test_jit_sort_joins() -> None:
        n = 200
        # Explicitly specify numpy dtype because of different defaults on Windows
        dfa = pd.DataFrame(
            {
                "a": np.random.randint(0, 100, n, dtype=np.int64),
                "b": np.arange(0, n, dtype=np.int64),
            }
        )
    
        n = 40
        dfb = pd.DataFrame(
            {
                "a": np.random.randint(0, 100, n, dtype=np.int64),
                "b": np.arange(0, n, dtype=np.int64),
            }
        )
        dfa_pl = pl.from_pandas(dfa).sort("a")
        dfb_pl = pl.from_pandas(dfb)
    
        for how in ["left", "inner"]:
            pd_result = dfa.merge(dfb, on="a", how=how)
            pd_result.columns = ["a", "b", "b_right"]
    
            # left key sorted right is not
            pl_result = dfa_pl.join(dfb_pl, on="a", how=how).sort(["a", "b"])
    
            a = pl.from_pandas(pd_result).with_columns(pl.all().cast(int)).sort(["a", "b"])
            assert_frame_equal(a, pl_result)
            assert pl_result["a"].flags["SORTED_ASC"]
    
            # left key sorted right is not
            pd_result = dfb.merge(dfa, on="a", how=how)
            pd_result.columns = ["a", "b", "b_right"]
>           pl_result = dfb_pl.join(dfa_pl, on="a", how=how).sort(["a", "b"])

tests/unit/test_joins.py:571: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
polars/internals/dataframe/frame.py:4311: in join
    .collect(no_optimization=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <polars.LazyFrame object at 0x7F671DFD7D90>

    def collect(
        self,
        *,
        type_coercion: bool = True,
        predicate_pushdown: bool = True,
        projection_pushdown: bool = True,
        simplify_expression: bool = True,
        no_optimization: bool = False,
        slice_pushdown: bool = True,
        common_subplan_elimination: bool = True,
        streaming: bool = False,
    ) -> pli.DataFrame:
        """
        Collect into a DataFrame.
    
        Note: use :func:`fetch` if you want to run your query on the first `n` rows
        only. This can be a huge time saver in debugging queries.
    
        Parameters
        ----------
        type_coercion
            Do type coercion optimization.
        predicate_pushdown
            Do predicate pushdown optimization.
        projection_pushdown
            Do projection pushdown optimization.
        simplify_expression
            Run simplify expressions optimization.
        no_optimization
            Turn off (certain) optimizations.
        slice_pushdown
            Slice pushdown optimization.
        common_subplan_elimination
            Will try to cache branching subplans that occur on self-joins or unions.
        streaming
            Run parts of the query in a streaming fashion (this is in an alpha state)
    
        Returns
        -------
        DataFrame
    
        Examples
        --------
        >>> df = pl.DataFrame(
        ...     {
        ...         "a": ["a", "b", "a", "b", "b", "c"],
        ...         "b": [1, 2, 3, 4, 5, 6],
        ...         "c": [6, 5, 4, 3, 2, 1],
        ...     }
        ... ).lazy()
        >>> df.groupby("a", maintain_order=True).agg(pl.all().sum()).collect()
        shape: (3, 3)
        ┌─────┬─────┬─────┐
        │ a   ┆ b   ┆ c   │
        │ --- ┆ --- ┆ --- │
        │ str ┆ i64 ┆ i64 │
        ╞═════╪═════╪═════╡
        │ a   ┆ 4   ┆ 10  │
        │ b   ┆ 11  ┆ 10  │
        │ c   ┆ 6   ┆ 1   │
        └─────┴─────┴─────┘
    
        """
        if no_optimization:
            predicate_pushdown = False
            projection_pushdown = False
            slice_pushdown = False
            common_subplan_elimination = False
    
        if streaming:
            common_subplan_elimination = False
    
        ldf = self._ldf.optimization_toggle(
            type_coercion,
            predicate_pushdown,
            projection_pushdown,
            simplify_expression,
            slice_pushdown,
            common_subplan_elimination,
            streaming,
        )
>       return pli.wrap_df(ldf.collect())
E       pyo3_runtime.PanicException: assertion failed: idx[0] <= idx[idx.len() - 1]

polars/internals/lazyframe/frame.py:1147: PanicException
---------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------
thread '<unnamed>' panicked at 'assertion failed: idx[0] <= idx[idx.len() - 1]', /home/stijn/code/polars/polars/polars-core/src/frame/mod.rs:3227:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
============================================================== short test summary info ===============================================================
FAILED tests/unit/test_joins.py::test_jit_sort_joins - pyo3_runtime.PanicException: assertion failed: idx[0] <= idx[idx.len() - 1]
```

### Reproducible example

```python
pytest -n 12
```


### Expected behavior

Test should pass.

### Installed versions

<details>

```
Master branch
```

</details>

--- a/polars/polars-core/src/frame/hash_join/mod.rs
+++ b/polars/polars-core/src/frame/hash_join/mod.rs
@@ -357,13 +357,13 @@ impl DataFrame {
         if left_join && join_tuples.len() == self.height() {
             self.clone()
         } else {
+            // left join tuples are always in ascending order
             let sorted = if left_join || sorted {
                 IsSorted::Ascending
             } else {
                 IsSorted::Not
             };
 
-            // left join tuples are always in ascending order
             self._take_unchecked_slice2(join_tuples, true, sorted)
         }
     }

--- a/polars/polars-core/src/frame/hash_join/sort_merge.rs
+++ b/polars/polars-core/src/frame/hash_join/sort_merge.rs
@@ -188,6 +188,10 @@ pub fn _sort_or_hash_inner(
     s_right: &Series,
     verbose: bool,
 ) -> ((Vec<IdxSize>, Vec<IdxSize>), bool) {
+    // We check if keys are sorted.
+    // - If they are we can do a sorted merge join
+    // If one of the keys is not, it can still be faster to sort that key and use
+    // the `arg_sort` indices to revert the sort once the join keys are determined.
     let size_factor_rhs = s_right.len() as f32 / s_left.len() as f32;
     let size_factor_lhs = s_left.len() as f32 / s_right.len() as f32;
     let size_factor_acceptable = std::env::var("POLARS_JOIN_SORT_FACTOR")
@@ -247,7 +251,8 @@ pub fn _sort_or_hash_inner(
                 });
             });
 
-            ((left, right), true)
+            // set sorted to `false` as we reverse sorted the left key.
+            ((left, right), false)
         }
         _ => s_left.hash_join_inner(s_right),
     }

